File: app/src/main/java/com/khmelenko/lab/varis/network/retrofit/travis/TravisApiService.java
package com.khmelenko.lab.varis.network.retrofit.travis;


import com.khmelenko.lab.varis.network.request.AccessTokenRequest;
import com.khmelenko.lab.varis.network.response.AccessToken;
import com.khmelenko.lab.varis.network.response.Branch;
import com.khmelenko.lab.varis.network.response.Branches;
import com.khmelenko.lab.varis.network.response.BuildDetails;
import com.khmelenko.lab.varis.network.response.Logs;
import com.khmelenko.lab.varis.network.response.Repo;
import com.khmelenko.lab.varis.network.response.BuildHistory;
import com.khmelenko.lab.varis.network.response.Request;
import com.khmelenko.lab.varis.network.response.Requests;
import com.khmelenko.lab.varis.network.response.User;

import java.util.List;

import retrofit.http.Body;
import retrofit.http.EncodedPath;
import retrofit.http.GET;
import retrofit.http.POST;
import retrofit.http.Path;
import retrofit.http.Query;
import retrofit.mime.TypedOutput;

/**
 * Defines an interface for the REST service
 *
 * @author Dmytro Khmelenko
 */
public interface TravisApiService {

    @POST("/auth/github")
    AccessToken auth(@Body AccessTokenRequest accessToken);

    // repositories
    @GET("/repos")
    List<Repo> getRepos();

    @GET("/repos")
    List<Repo> getRepos(@Query("search") String search);

    @GET("/repos/{repositoryId}")
    Repo getRepo(@Path("repositoryId") long repositoryId);

    @GET("/repos/{repositorySlug}")
    Repo getRepo(@EncodedPath("repositorySlug") String repositorySlug);

    @GET("/repos")
    List<Repo> getUserRepos(@Query("member") String userName);


    // branches
    @GET("/repos/{repositoryId}/branches")
    Branches getBranches(@Path("repositoryId") long repositoryId);

    @GET("/repos/{repositorySlug}/branches")
    Branches getBranches(@EncodedPath("repositorySlug") String repositorySlug);

    @GET("/repos/{repositoryId}/branches/{branch}")
    Branch getBranch(@Path("repositoryId") long repositoryId, @Path("branch") String branch);

    @GET("/repos/{repositorySlug}/branches/{branch}")
    Branch getBranch(@EncodedPath("repositorySlug") String repositorySlug, @Path("branch") String branch);


    // builds
    @GET("/repos/{repositoryId}/builds")
    BuildHistory getBuilds(@Path("repositoryId") long repositoryId);

    @GET("/repos/{repositorySlug}/builds")
    BuildHistory getBuilds(@EncodedPath("repositorySlug") String repositorySlug);

    @GET("/repos/{repositoryId}/builds?event_type=pull_request")
    BuildHistory getPullRequestBuilds(@Path("repositoryId") long repositoryId);

    @GET("/repos/{repositorySlug}/builds?event_type=pull_request")
    BuildHistory getPullRequestBuilds(@EncodedPath("repositorySlug") String repositorySlug);

    @GET("/repos/{repositoryId}/builds/{buildId}")
    BuildDetails getBuild(@Path("repositoryId") long repositoryId, @Path("buildId") long buildId);

    @GET("/repos/{repositorySlug}/builds/{buildId}")
    BuildDetails getBuild(@EncodedPath("repositorySlug") String repositorySlug, @Path("buildId") long buildId);

    @POST("/builds/{buildId}/cancel")
    Object cancelBuild(@Path("buildId") long buildId, @Body TypedOutput emptyBody);

    @POST("/builds/{buildId}/restart")
    Object restartBuild(@Path("buildId") long buildId, @Body TypedOutput emptyBody);


    // requests
    @GET("/requests/{requestId}")
    Request getRequest(@Path("requestId") long requestId);

    @GET("/requests")
    Requests getRequests(@Query("repository_id") long repositoryId);

    @GET("/requests")
    Requests getRequests(@Query("slug") String repositorySlug);


    // logs
    @GET("/jobs/{jobId}/log")
    String getLogRaw(@Path("jobId") long jobId);

    @GET("/logs/{logId}")
    Logs getLogs(@Path("logId") long logId);


    // users
    @GET("/users")
    User getUser();
}

File: app-v3/src/main/java/com/khmelenko/lab/varis/network/retrofit/travis/TravisApiService.kt
package com.khmelenko.lab.varis.network.retrofit.travis

import com.khmelenko.lab.varis.network.request.AccessTokenRequest
import com.khmelenko.lab.varis.network.response.AccessToken
import com.khmelenko.lab.varis.network.response.Branch
import com.khmelenko.lab.varis.network.response.Branches
import com.khmelenko.lab.varis.network.response.BuildDetails
import com.khmelenko.lab.varis.network.response.BuildHistory
import com.khmelenko.lab.varis.network.response.Logs
import com.khmelenko.lab.varis.network.response.Repo
import com.khmelenko.lab.varis.network.response.Request
import com.khmelenko.lab.varis.network.response.Requests
import com.khmelenko.lab.varis.network.response.User

import io.reactivex.Single
import okhttp3.RequestBody
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query


interface TravisApiService {

    // repositories
    @get:GET("/repos")
    val repos: Single<List<Repo>>


    // users
    @get:GET("/users")
    val user: Single<User>

    @POST("/auth/github")
    fun auth(@Body accessToken: AccessTokenRequest): Single<AccessToken>

    @GET("/repos")
    fun getRepos(@Query("search") search: String): Single<List<Repo>>

    @GET("/repos/{repositoryId}")
    fun getRepo(@Path("repositoryId") repositoryId: Long): Single<Repo>

    @GET("/repos/{repositorySlug}")
    fun getRepo(@Path(value = "repositorySlug", encoded = true) repositorySlug: String): Single<Repo>

    @GET("/repos")
    fun getUserRepos(@Query("member") userName: String): Single<List<Repo>>


    // branches
    @GET("/repos/{repositoryId}/branches")
    fun getBranches(@Path("repositoryId") repositoryId: Long): Single<Branches>

    @GET("/repos/{repositorySlug}/branches")
    fun getBranches(@Path(value = "repositorySlug", encoded = true) repositorySlug: String): Single<Branches>

    @GET("/repos/{repositoryId}/branches/{branch}")
    fun getBranch(@Path("repositoryId") repositoryId: Long, @Path("branch") branch: String): Single<Branch>

    @GET("/repos/{repositorySlug}/branches/{branch}")
    fun getBranch(@Path(value = "repositorySlug", encoded = true) repositorySlug: String, @Path("branch") branch: String): Single<Branch>


    // builds
    @GET("/repos/{repositoryId}/builds")
    fun getBuilds(@Path("repositoryId") repositoryId: Long): Single<BuildHistory>

    @GET("/repos/{repositorySlug}/builds")
    fun getBuilds(@Path(value = "repositorySlug", encoded = true) repositorySlug: String): Single<BuildHistory>

    @GET("/repos/{repositoryId}/builds?event_type=pull_request")
    fun getPullRequestBuilds(@Path("repositoryId") repositoryId: Long): Single<BuildHistory>

    @GET("/repos/{repositorySlug}/builds?event_type=pull_request")
    fun getPullRequestBuilds(@Path(value = "repositorySlug", encoded = true) repositorySlug: String): Single<BuildHistory>

    @GET("/repos/{repositoryId}/builds/{buildId}")
    fun getBuild(@Path("repositoryId") repositoryId: Long, @Path("buildId") buildId: Long): Single<BuildDetails>

    @GET("/repos/{repositorySlug}/builds/{buildId}")
    fun getBuild(@Path(value = "repositorySlug", encoded = true) repositorySlug: String, @Path("buildId") buildId: Long): Single<BuildDetails>

    @POST("/builds/{buildId}/cancel")
    fun cancelBuild(@Path("buildId") buildId: Long, @Body emptyBody: RequestBody): Single<Any>

    @POST("/builds/{buildId}/restart")
    fun restartBuild(@Path("buildId") buildId: Long, @Body emptyBody: RequestBody): Single<Any>


    // requests
    @GET("/requests/{requestId}")
    fun getRequest(@Path("requestId") requestId: Long): Single<Request>

    @GET("/requests")
    fun getRequests(@Query("repository_id") repositoryId: Long): Single<Requests>

    @GET("/requests")
    fun getRequests(@Query("slug") repositorySlug: String): Single<Requests>


    // logs
    @GET("/jobs/{jobId}/log")
    fun getLogRaw(@Path("jobId") jobId: Long): Single<String>

    @GET("/logs/{logId}")
    fun getLogs(@Path("logId") logId: Long): Single<Logs>
}
